Mathlib/Tactic/GeneralizeProofs.lean

```diff
@@ -1,118 +1,522 @@
 /-
 Copyright (c) 2022 Alex J. Best. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Alex J. Best
+Authors: Alex J. Best, Kyle Miller
 -/
-import Lean.Meta.AbstractNestedProofs
-import Lean.Meta.Tactic.Generalize
+import Lean.Elab.Tactic.Config
 import Lean.Elab.Tactic.Location
 import Mathlib.Lean.Expr.Basic
 
 /-!
 # The `generalize_proofs` tactic
 
-Generalize any proofs occurring in the goal or in chosen hypotheses, replacing them by
-named hypotheses so that they can be referred to later in the proof easily.
-Commonly useful when dealing with functions like `Classical.choose` that produce data from proofs.
+Generalize any proofs occurring in the goal or in chosen hypotheses,
+replacing them by local hypotheses.
+When these hypotheses are named, this makes it easy to refer to these proofs later in a proof,
+commonly useful when dealing with functions like `Classical.choose` that produce data from proofs.
+It is also useful to eliminate proof terms to handle issues with dependent types.
 
 For example:
 ```lean
-example : List.nthLe [1, 2] 1 dec_trivial = 2 := by
-  -- ⊢ [1, 2].nthLe 1 _ = 2
-  generalize_proofs h,
+example : List.nthLe [1, 2] 1 (by simp) = 2 := by
+  -- ⊢ [1, 2].nthLe 1 ⋯ = 2
+  generalize_proofs h
   -- h : 1 < [1, 2].length
   -- ⊢ [1, 2].nthLe 1 h = 2
 ```
+
+The tactic is similar in spirit to `Lean.Meta.AbstractNestedProofs` in core.
+One difference is that it the tactic tries to propagate expected types so that
+we get `1 < [1, 2].length` in the above example rather than `1 < Nat.succ 1`.
 -/
 
-namespace Mathlib.Tactic.GeneralizeProofs
+namespace Mathlib.Tactic
 open Lean Meta Elab Parser.Tactic Elab.Tactic
 
-/- The following set up are the visit function are based on the file
-Lean.Meta.AbstractNestedProofs in core -/
+initialize registerTraceClass `Tactic.generalize_proofs
+
+namespace GeneralizeProofs
+
+/--
+Configuration for the `generalize_proofs` tactic.
+-/
+structure Config where
+  /-- The maximum recursion depth when generalizing proofs.
+  When `maxDepth > 0`, then proofs are generalized from the types of the generalized proofs too. -/
+  maxDepth : Nat := 8
+  /-- When `abstract` is `true`, then the tactic will create universally quantified proofs
+  to account for bound variables.
+  When it is `false` then such proofs are left alone. -/
+  abstract : Bool := true
+  /-- (Debugging) When `true`, enables consistency checks. -/
+  debug : Bool := true
```

Comment:
seems odd to have a debug arg true by default. Maybe its good to have this always on still? But in that case maybe it shouldn't be called debug!

Reply:
Thanks, I forgot to set it back to `false`. I had that set for the full mathlib build to make sure some invariants were actually holding.

In principle the consistency checks don't do anything, since they're just checking invariants, but in practice I had a lot of trouble making sure the invariants actually held!